Mathlib/Topology/Algebra/TopologicallyNilpotent.lean

```diff
@@ -0,0 +1,146 @@
+open scoped Topology
+
+/-- An element is topologically nilpotent if its powers converge to `0`. -/
+def IsTopologicallyNilpotent
+    {R : Type*} [MonoidWithZero R] [TopologicalSpace R] (a : R) : Prop :=
+  Tendsto (fun n : ℕ => a ^ n) atTop (𝓝 0)
+
+namespace IsTopologicallyNilpotent
+
+section MonoidWithZero
+
+variable {R S : Type*} [TopologicalSpace R] [MonoidWithZero R]
+  [MonoidWithZero S] [TopologicalSpace S]
+
+/-- The image of a topologically nilpotent element under a continuous morphism
+  is topologically nilpotent -/
+theorem map {F : Type*} [FunLike F R S] [MonoidWithZeroHomClass F R S]
+    {φ : F} (hφ : Continuous φ) {a : R} (ha : IsTopologicallyNilpotent a) :
+    IsTopologicallyNilpotent (φ a) := by
+  unfold IsTopologicallyNilpotent at ha ⊢
+  simp_rw [← map_pow]
+  exact (map_zero φ ▸  hφ.tendsto 0).comp ha
+
+/-- `0` is topologically nilpotent -/
+theorem zero :
+    IsTopologicallyNilpotent (0 : R) := tendsto_atTop_of_eventually_const (i₀ := 1)
+    (fun _ hi => by rw [zero_pow (Nat.ne_zero_iff_zero_lt.mpr hi)])
+
+end MonoidWithZero
+
+section Ring
+
+variable {R : Type*} [TopologicalSpace R] [Ring R]
+
+/-- If `a` and `b` commute and `a` is topologically nilpotent,
+  then `a * b` is topologically nilpotent. -/
+theorem mul_right_of_commute [IsLinearTopology Rᵐᵒᵖ R]
+    {a b : R} (ha : IsTopologicallyNilpotent a) (hab : Commute a b) :
+    IsTopologicallyNilpotent (a * b) := by
+  intro v hv
+  rw [(IsLinearTopology.hasBasis_submodule Rᵐᵒᵖ).mem_iff] at hv
+  rcases hv with ⟨I, I_mem_nhds, I_subset⟩
+  specialize ha I_mem_nhds
+  simp only [mem_map, mem_atTop_sets, ge_iff_le, Set.mem_preimage, SetLike.mem_coe] at ha ⊢
+  rcases ha with ⟨n, ha⟩
+  use n
+  intro m hm
+  rw [hab.mul_pow]
+  exact I_subset (I.smul_mem _ (ha m hm))
+
+/-- If `a` and `b` commute and `b` is topologically nilpotent,
+  then `a * b` is topologically nilpotent. -/
+ theorem mul_left_of_commute [IsLinearTopology R R] {a b : R}
+    (hb : IsTopologicallyNilpotent b) (hab : Commute a b) :
+    IsTopologicallyNilpotent (a * b) := by
+  intro v hv
+  rw [IsLinearTopology.hasBasis_ideal.mem_iff] at hv
+  rcases hv with ⟨I, I_mem_nhds, I_subset⟩
+  specialize hb I_mem_nhds
+  simp only [mem_map, mem_atTop_sets, ge_iff_le, Set.mem_preimage, SetLike.mem_coe] at hb ⊢
+  rcases hb with ⟨n, hb⟩
+  use n
+  intro m hm
+  rw [hab.mul_pow]
+  exact I_subset (I.mul_mem_left _ (hb m hm))
+
+/-- If `a` and `b` are topologically nilpotent and commute,
+  then `a + b` is topologically nilpotent. -/
+theorem add_of_commute [IsLinearTopology R R] [IsLinearTopology Rᵐᵒᵖ R]
+    {a b : R} (ha : IsTopologicallyNilpotent a) (hb : IsTopologicallyNilpotent b)
+    (h : Commute a b) :
+    IsTopologicallyNilpotent (a + b) := fun v ↦ by
+  rw [IsLinearTopology.hasBasis_twoSidedIdeal.mem_iff]
+  rintro ⟨I, I_mem_nhds, I_subset⟩
+  specialize ha I_mem_nhds
+  specialize hb I_mem_nhds
+  simp only [mem_map, mem_atTop_sets, ge_iff_le, Set.mem_preimage, SetLike.mem_coe] at ha hb ⊢
```

Comment:
I think you should be able to remove some of the boilerplate by using [Filter.HasBasis.tendsto_iff](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Order/Filter/Bases.html#Filter.HasBasis.tendsto_iff), applied to [IsLinearTopology.hasBasis_twoSidedIdeal](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Topology/Algebra/LinearTopology.html#IsLinearTopology.hasBasis_twoSidedIdeal) and [Filter.atTop_basis](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Order/Filter/AtTopBot.html#Filter.atTop_basis).

Reply:
I did not understand what you wanted me to do, I added a lemma and simplified the proof.